Reduce size of Unicode tables

Kmeakin opened the pull request:
Follow up to https://github.com/rust-lang/rust/pull/145027.
Shave a few bytes from the tables by:
* Removing ASCII characters from the sets: 31446 bytes to 31420 bytes
* Replacing `Cased` with `Titlecase_letter`: 31420 bytes to 31050 bytes
* Using `match` expressions for sufficiently small sets 31050 bytes to 30754 bytes

Kmeakin committed:
fix: Fix panic in `unicode-table-generator`

Commit 15acb0e0ec849453d9db928821cecc552bb51bfd introduced a panic when
running `./x run tools/unicode-table-generator`. Fix it by undoing one
of the refactors.

Kmeakin committed:
optimization: Don't include ASCII characters in Unicode tables

The ASCII subset of Unicode is fixed and will never change, so we don't
need to generate tables for it with every new Unicode version. This
saves a few bytes of static data and speeds up `char::is_control` and
`char::is_grapheme_extended` on ASCII inputs.

Since the table lookup functions exported from the `unicode` module will
give nonsensical errors on ASCII input (and in fact will panic in debug
mode), I had to add some private wrapper methods to `char` which check
for ASCII-ness first.

Kmeakin committed:
refactor: Include table sizes in comment at top of `unicode_data.rs`

To make changes in table size obvious from git diffs

Kmeakin committed:
refactor: Include size of case conversion tables

Include the sizes of the `to_lowercase` and `to_uppercase` tables in the
total size calculations.

Kmeakin committed:
optimization: Use `match` for small sets

If the number of codepoint ranges in a set is sufficiently small, it may
be better to simply use a `match` expression rather than a lookup table.
The instructions to implement the `match` may be slightly bigger than
the table that it replaced (hard to predict, depends on architecture and
whatever optimzations LLVM applies), but in return we elimate the lookup
tables and avoid the slower binary search.

Kmeakin committed:
refactor: rewrite `ranges_from_set`

The `merge_ranges` function was very complicated and hard to understand.
Forunately, we can use `slice::chunk_by` to achieve the same thing.

Kmeakin committed:
refactor: `generate_tests`

Rewrite `generate_tests` to be more idiomatic.

Kmeakin committed:
refactor: Add tests for case conversions

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16864555062/job/47769623642) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47769623642) [(plain)](https://github.com/rust-lang/rust/commit/c18f085be8d2d08bd0612c9c59d2b2a40ef86066/checks/47769623642/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- str::to_lowercase stdout ----
---- str::to_lowercase stderr ----

thread 'main' (32284) panicked at /checkout/library/core/src/unicode/unicode_data.rs:321:9:
assertion failed: !c.is_ascii()
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: alloc::str::<impl str>::to_lowercase
---
    str::to_lowercase

test result: FAILED. 825 passed; 1 failed; 26 ignored; 0 measured; 0 filtered out; finished in 1.05s

error: test failed, to rerun pass `-p alloctests --test alloctests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.91.0-nightly\t(eb2a22cec\t2025-08-10)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-19/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "aarch64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:08:08
  local time: Sun Aug 10 17:57:18 UTC 2025
  network time: Sun, 10 Aug 2025 17:57:18 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Kmeakin committed:
optimization: Eliminate `Cased` table

`Cased` is a derived property - it is the union of the `Lowercase`
property, the `Uppercase` property, and the `Titlecase_Letter` generaral
category. We already have lookup tables for `Lowercase` and `Uppercase`,
and `Titlecase_Letter` is very small. So instead of duplicating a lookup
table for `Cased`, just test each of those properties in turn.

This probably will be slower than the old approach, but it is not a
public API: it is only used in `string::to_lower` when deciding when a
Greek "sigma" should be mapped to `ς` or to `σ`. This is a very rare
case, so should not be performance sensitive.

joshtriplett commented:
@Kmeakin For that last change to use `match`, do you have any way to check how the code size change balances with the table size change? Could you build a standalone program that includes the tables and code, calls all the functions once to ensure they're used, and check the *total* size of that program, code and data?

Kmeakin commented:
> @Kmeakin For that last change to use `match`, do you have any way to check how the code size change balances with the table size change? Could you build a standalone program that includes the tables and code, calls all the functions once to ensure they're used, and check the _total_ size of that program, code and data?

https://godbolt.org/z/ef5ExG5Eo
`is_whitespace` grows slightly larger, but is offset by getting rid of the 256 bytes of static data
`is_control` gets slightly worse: LLVM doesn't seem to realise that `matches!(c, 0x00..=0x1f | 0x7f | 0x80..=0x9f)` can be optimized to `matches!(c, 0x00..=0x1f | 0x7f..=0x9f)`

okaneco commented:
This was the PR that added a lookup table for `is_whitespace` https://github.com/rust-lang/rust/pull/99487

The trade-off is execution speed versus the 256 byte table. Unfortunately, there aren't any benches in tree but the author of that PR provided a repo they instrumented with criterion. Some of those examples could be included as well as benching on the current corpora.

I suspect the performance of the match still is a bit slower than the current table-based implementation for `is_whitespace`.

Kmeakin committed:
refactor: Hard-code `char::is_control`

According to
https://www.unicode.org/policies/stability_policy.html#Property_Value,
the set of codepoints in `Cc` will never change. So we can hard-code
the patterns to match against instead of using a table.

Kmeakin commented:
> * `is_control` gets slightly worse: LLVM doesn't seem to realise that `matches!(c, 0x00..=0x1f | 0x7f | 0x80..=0x9f)` can be optimized to `matches!(c, 0x00..=0x1f | 0x7f..=0x9f)`

Actually, [`Cc` is guaranteed not to change in the future](https://www.unicode.org/policies/stability_policy.html#Property_Value), so we can just hardcode it :)

rust-timer commented:
Finished benchmarking commit (ce72b92f7cb12f5a2f1362968dbe742897e3a4d7): [comparison URL](https://perf.rust-lang.org/compare.html?start=350d0ef0ec0493e6d21cfb265cb8211a0e74d766&end=ce72b92f7cb12f5a2f1362968dbe742897e3a4d7&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=350d0ef0ec0493e6d21cfb265cb8211a0e74d766&end=ce72b92f7cb12f5a2f1362968dbe742897e3a4d7&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.3%, 1.0%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.9%  | [0.9%, 0.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.2%, -0.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-1.6%, -0.2%] | 3     |
| All ❌✅ (primary)                 | 0.2%  | [-0.2%, 1.0%]  | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=350d0ef0ec0493e6d21cfb265cb8211a0e74d766&end=ce72b92f7cb12f5a2f1362968dbe742897e3a4d7&stat=max-rss)
<details>
<summary>Results (primary -2.9%, secondary -2.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.9% | [-2.9%, -2.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.4%, -2.4%] | 1     |
| All ❌✅ (primary)                 | -2.9% | [-2.9%, -2.9%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=350d0ef0ec0493e6d21cfb265cb8211a0e74d766&end=ce72b92f7cb12f5a2f1362968dbe742897e3a4d7&stat=cycles:u)
<details>
<summary>Results (secondary -2.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-3.4%, -1.3%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=350d0ef0ec0493e6d21cfb265cb8211a0e74d766&end=ce72b92f7cb12f5a2f1362968dbe742897e3a4d7&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary -0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.5%]   | 17    |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.1%, 1.1%]   | 7     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.3%, -0.0%] | 16    |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.0%] | 43    |
| All ❌✅ (primary)                 | -0.0% | [-0.3%, 0.5%]  | 33    |
</details>

**Bootstrap**: 468.291s -> 469.372s (0.23%)
**Artifact size**: 377.36 MiB -> 377.29 MiB (-0.02%)

Kmeakin commented:
This PR has accumulated quite a few separate optimizations:
* filtering out ASCII characters
* hard-coding `char::is_control`
* using `match` for `char::is_whitespace`
* replacing `cased` with `lowercase || uppercase || titlecase`

I think it would be best to split into separate PRs so the impact of each optimization can be evaluated separately. @scottmcm , @joshtriplett is that ok with you?
